plutus-tx/src/PlutusTx/BuiltinList.hs

```diff
@@ -69,21 +71,15 @@ all p = go
 
 {-| Get the element at a given index.
 
-This function is partial and takes linear time.
+This function throws an error if the index is negative or
+larger than the length of the list.
 -}
 (!!) :: forall a. BuiltinList a -> Integer -> a
-(!!) xs0 i0
-  | i0 `B.lessThanInteger` 0 = traceError builtinListNegativeIndexError
-  | otherwise = go xs0 i0
- where
-  go :: BuiltinList a -> Integer -> a
-  go xs i =
-    B.caseList
-      (\_ -> traceError builtinListIndexTooLargeError)
-      ( \y ys _ ->
-          if i `B.equalsInteger` 0
-            then y
-            else go ys (B.subtractInteger i 1)
-      )
-      xs
-      ()
+(!!) xs i
+  | i `B.lessThanInteger` 0 = traceError builtinListNegativeIndexError
+  | otherwise =
+      B.caseList
+        (\_ann -> traceError builtinListIndexTooLargeError)
+        (\y _rest _ann -> y)
+        (BI.drop i xs)
+        ()
```

Comment:
This is outside of the scope of this PR, but these conditions to throw specific trace messages make function less efficient. And when script is compiled with traces turned off, these are checks are with no purpose really. 

For example, this function will be much more efficient as `BI.head $ BI.drop i xs` when script is compiled with traces turned off.

Reply:
In fact, my first implementation was exactly what you wrote `BI.head (BI.drop i xs)` but then I saw it as a breaking backward compatibility in the negative case and decided to preserve the behavior as is (at least for this PR).

Your observation is legit! I propose you bring it up in a more general form "How much of runtime checks we want to use in Plinth std lib" for us to come up with a guiding principle (rule of 👍🏼 ), if possible.

Reply:
Ideally, we want to have two implementation that we can choose depending on rather we are compiling to tracing script or not as we have no use whatsoever for these runtime checks for non-tracing scripts.

This is kind of tricky since this means we either 
- have two separate implementations of same functions 
- or, have some way to signal(Perhaps implicit parameter or nullary typeclass--I don't like either :/ ) implementation to automatically pick what implementation to use.

Both option does feel rather janky at the moment.